universal-application-tool-0.0.1/app/services/applicant/predicate/JsonPathPredicateGenerator.java

```diff
@@ -1,20 +1,45 @@
 package services.applicant.predicate;
 
-import services.Path;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import services.applicant.exception.InvalidPredicateException;
+import services.applicant.question.ApplicantQuestion;
 import services.program.predicate.LeafOperationExpressionNode;
 
+/** Generates {@link JsonPathPredicate}s based on the current applicant filling out the program. */
 public class JsonPathPredicateGenerator {
 
+  private final ImmutableMap<Long, ApplicantQuestion> questionsById;
+
+  public JsonPathPredicateGenerator(ImmutableList<ApplicantQuestion> programQuestions) {
```

Comment:
(idea): if this object was constructed with the `Block`'s `RepeatedEntity` and the `QuestionDefinition`s, then the `Path` for the predicate could be generated by `repeatedEntity.getContextualizedPath().join(questionDefinition.getFormattedName()).

We'd have to "walk up" the `RepeatedEntity` chain to find the repeated entity whose enumerator id matches the question's enumerator id.

Reply:
Thanks Kendrick! I added this logic and unit tests - let me know if there are other tests you think would be helpful here